### ISSUE TITLE
Some question about profiling gevent based web application

### ISSUE THREAD

**Issue by monsterxx03** [uber/pyflame#64]
My application is a web application hosted by gunicorn, using gevent as async worker.

gunicorn pre-forked two processes, every one is a gevent worker (single thread). My attachment graph is profiling for one worker.

From the doc, since pyflame can't profile IO and c code, I think my left IDLE part should be related to it.

The middle part is my real business logic related code.

For  the right long part,  all in gevent.threadpool:_worker (task = task_queue.get()), I'm not familiar with gevent internal, is it normal?  seems it's gevent main greenlet waiting for available tasks, My guess is because gevent  is blocked by some CPU heavy code. But how can it be profiled with my real application code on the same level (If my assumption is right, the right part and middle part should be mixed together)

If only look at  the middle part (my business logic code), does it only means CPU usage of my call stack? For example I have a very complex function, it will do both heavy CPU work  and call a slow rpc service (waiting for IO). From pyflame, the waiting for IO part should be in IDLE, so my rpc part on flame graph will be narrow, is my understanding right?

Thanks for any advice.

![screen shot 2017-04-26 at 10 09 45 am](https://cloud.githubusercontent.com/assets/813489/25415354/b307cf08-2a69-11e7-8055-7fcce22fe486.png)

**Comment by workingenius** [comment]
similar situation. searching for an answer.

**Comment by jamespic** [comment]
This looks like it *might* be caused by #62, which is now fixed in master. Could you have a go with the latest build?

If not, could you post a snippet that'll reproduce the issue, and I'll see if I can get to the bottom of it. Greenlets can wreak havoc on stack inspection, so it might just be that, but I'll see if I can make sense of it either way.

**Comment by monsterxx03** [comment]
Thanks, I tested with pyflame 1.3.1.  

The right gevent part disappeared.  Now I only have IDLE part and my logic part.

And so as my last part in the question. Did my logic part  only means CPU usage of my call stack?

If my python function  is waiting for IO (eg:  calling rpc server via socket), if pyflame do snapshot on it, the sample will be in IDLE? 

I did a test, a simple http server, just sleeping for long time:

```
import gevent
from gevent import pywsgi

def hello_world(env, start_response):
    gevent.sleep(1000)
    start_response('200 OK', [('Content-Type', 'text/html')])
    return ["<b>hello world</b>"]

pywsgi.WSGIServer(('127.0.0.1', 8001), hello_world).serve_forever()
```

A  simple client to call this server:

```
import requests
print requests.get('http://127.0.0.1:8001')
```

Then use pyflame to profile client, 100% sample in IDLE. 

Seems my assumption is right. I original guess is pyflame will see something like `sock.recv`

**Comment by jamespic** [comment]
@monsterxx03 As a rough approximation, yes. With the latest version of PyFlame, unless you've run it in `--threads` mode, IO should show up as `(idle)`.

More precisely, any time when it's not executing Python code is listed as `(idle)`, which often means IO, although some native code such as numerical libraries would show up as `(idle)`.

Even more precisely still, it's about the global interpreter lock. CPython protects the internals of the interpreter with a single lock, the GIL, and anything that needs to interact with Python objects has to acquire this lock for safety (and code that doesn't currently need it, because it's waiting for stuff like IO or pure C calculations, can release it until it needs it). By default, PyFlame will only profile the thread that holds the GIL, and will report time as `(idle)` if no thread currently does.

This issue was caused by some refactoring we did to add the `--threads` flag, which allows PyFlame to profile all threads, not just the one with the GIL. When we first implemented this, there was a bug, #62, which meant that when you were running in the default (not `--threads`) mode, it would only profile *when* something held the GIL, but it would profile *all* threads (including ones waiting for IO, or otherwise not holding the GIL), not just the one with the GIL.

So in your case, the gevent code that was waiting for IO would be spuriously added to the timings whenever something held the GIL.

Hopefully this should be sorted now, but if you can find a test case that still triggers it, I'll take a look.

**Comment by monsterxx03** [comment]
Thanks so much, much clear now.

The gevent problem has been fixed in my case, I'll close this issue.
